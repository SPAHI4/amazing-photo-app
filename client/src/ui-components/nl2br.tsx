import React from 'react';

const newLineRegex = /(\r\n|\r|\n)/g;
const breakLine = /(\r\n|\r|\n)/;

interface Nl2BrProps {
  children: string;
  breakLimit?: number;
}

export function Nl2Br({ children: str, breakLimit = 2 }: Nl2BrProps): React.ReactNode {
  const result = str.split(newLineRegex);
  const output = [];

  let consecutiveBreaks = 0;

  for (let i = 0; i < result.length; i += 1) {
    if (result[i].trim() === '') {
      if (consecutiveBreaks < breakLimit + 1) {
        output.push(result[i]);
        consecutiveBreaks += 1;
      }
    } else {
      consecutiveBreaks = 0;
      output.push(result[i]);
    }
  }

  return output.map((line, index) => {
    if (line.match(breakLine)) {
      // eslint-disable-next-line react/no-array-index-key
      return <br key={index} />;
    }
    return line;
  });
}
