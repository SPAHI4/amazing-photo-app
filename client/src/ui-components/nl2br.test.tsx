import { render } from '@testing-library/react';

import { Nl2Br } from './nl2br';

describe('Nl2Br', () => {
  it('should render', () => {
    const { container } = render(<Nl2Br>foo</Nl2Br>);

    expect(container).toMatchSnapshot();
  });

  it('should render with breakLimit', () => {
    const { container } = render(
      <Nl2Br breakLimit={2}>{`
      foo
      
      
      
      
      
      bar
    `}</Nl2Br>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render with multiple new lines', () => {
    const { container } = render(
      <Nl2Br>
        {`foo bar
        
        
        
        foo bar
        
        foo bar
        
        `}
      </Nl2Br>,
    );

    expect(container).toMatchSnapshot();
  });
});
