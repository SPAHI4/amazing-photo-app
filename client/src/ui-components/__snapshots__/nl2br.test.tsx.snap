// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nl2Br should render 1`] = `
<div>
  foo
</div>
`;

exports[`Nl2Br should render with breakLimit 1`] = `
<div>
  <br />
        foo
  <br />
        
  <br />
        bar
  <br />
      
</div>
`;

exports[`Nl2Br should render with multiple new lines 1`] = `
<div>
  foo bar
  <br />
          
  <br />
          foo bar
  <br />
          
  <br />
          foo bar
  <br />
          
  <br />
</div>
`;
